packages/connection/src/Manager.php

```diff
@@ -145,15 +602,412 @@ public function verify_signature() {
 	 * WordPress.com.
 	 *
+	 * @param String $api_endpoint (optional) an API endpoint to use, defaults to 'register'.
 	 * @return Integer zero on success, or a bitmask on failure.
 	 */
-	public function register() {
-		return 0;
+	public function register( $api_endpoint = 'register' ) {
+
+		add_action( 'pre_update_jetpack_option_register', array( '\Jetpack_Options', 'delete_option' ) );
+		$secrets = $this->generate_secrets( 'register', get_current_user_id(), 600 );
+
+		if (
+			empty( $secrets['secret_1'] ) ||
+			empty( $secrets['secret_2'] ) ||
+			empty( $secrets['exp'] )
+		) {
+			return new \WP_Error( 'missing_secrets' );
+		}
+
+		// Better to try (and fail) to set a higher timeout than this system
+		// supports than to have register fail for more users than it should.
+		$timeout = $this->set_min_time_limit( 60 ) / 2;
+
+		$gmt_offset = get_option( 'gmt_offset' );
+		if ( ! $gmt_offset ) {
+			$gmt_offset = 0;
+		}
+
+		$stats_options = get_option( 'stats_options' );
+		$stats_id      = isset( $stats_options['blog_id'] )
+			? $stats_options['blog_id']
+			: null;
+
+		/**
+		 * Filters the request body for additional property addition.
+		 *
+		 * @since 7.6.0
+		 *
+		 * @param Array $post_data request data.
+		 * @param Array $token_data token data.
+		 */
+		$body = apply_filters(
+			'jetpack_register_request_body',
+			array(
+				'siteurl'         => site_url(),
+				'home'            => home_url(),
+				'gmt_offset'      => $gmt_offset,
+				'timezone_string' => (string) get_option( 'timezone_string' ),
+				'site_name'       => (string) get_option( 'blogname' ),
+				'secret_1'        => $secrets['secret_1'],
+				'secret_2'        => $secrets['secret_2'],
+				'site_lang'       => get_locale(),
+				'timeout'         => $timeout,
+				'stats_id'        => $stats_id,
+				'state'           => get_current_user_id(),
+				'site_created'    => $this->get_assumed_site_creation_date(),
+				'jetpack_version' => Constants::get_constant( 'JETPACK__VERSION' ),
+			)
+		);
+
+		$args = array(
+			'method'  => 'POST',
+			'body'    => $body,
+			'headers' => array(
+				'Accept' => 'application/json',
+			),
+			'timeout' => $timeout,
+		);
+
+		$args['body'] = $this->apply_activation_source_to_args( $args['body'] );
+
+		// TODO: fix URLs for bad hosts.
+		$response = Client::_wp_remote_request(
+			$this->api_url( $api_endpoint ),
+			$args,
+			true
+		);
+
+		// Make sure the response is valid and does not contain any Jetpack errors.
+		$registration_details = $this->validate_remote_register_response( $response );
+
+		if ( is_wp_error( $registration_details ) ) {
+			return $registration_details;
+		} elseif ( ! $registration_details ) {
+			return new \WP_Error(
+				'unknown_error',
+				'',
+				wp_remote_retrieve_response_code( $response )
+			);
+		}
+
+		if ( empty( $registration_details->jetpack_secret ) || ! is_string( $registration_details->jetpack_secret ) ) {
+			return new \WP_Error(
+				'jetpack_secret',
+				'',
```

Comment:
Maybe we could specify more information about the error? Like `A Jetpack secret is required for validating registration` or something.

Reply:
Added an error message here.